testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestReadUniontype.java

```diff
@@ -137,6 +139,79 @@ public void testReadUniontype(String storageFormat)
         }
     }
 
+    @Test(dataProvider = "storage_formats", groups = SMOKE)
+    public void testReadUniontypeWithDereference(String storageFormat)
+    {
+        // According to testing results, the Hive INSERT queries here only work in Hive 1.2
+        if (getHiveVersionMajor() != 1 || getHiveVersionMinor() != 2) {
+            throw new SkipException("This test can only be run with Hive 1.2 (default config)");
+        }
+        String tableUnionDereference = "test_union_dereference" + randomNameSuffix();
+
+        switch (storageFormat) {
+            case "AVRO":
+                testAvroUnionTypeDereference(tableUnionDereference);
+                break;
+            case "ORC":
+                testORCUnionTypeDereference(tableUnionDereference);
+                break;
+            default:
+                throw new UnsupportedOperationException("Unsupported table format.");
+        }
+    }
+
+    private void testAvroUnionTypeDereference(String tableName)
+    {
+        /*
+         * On hive 1.2, when create AVRO table with nested format, such as: uniontype<struct<unionlevel1:uniontype<string,int>>>
+         * will get result in: struct<unionlevel1:uniontype<string,int>
+         */
+        onHive().executeQuery(format(
+                "CREATE TABLE %s (unionLevel0 UNIONTYPE<" +
+                        "INT, STRING>)" +
+                        "STORED AS %s",
+                tableName,
+                "AVRO"));
+
+        onHive().executeQuery(format(
+                "INSERT INTO TABLE %s " +
+                        "SELECT create_union(0, 321, 'row1') " +
+                        "UNION ALL " +
+                        "SELECT create_union(1, 55, 'row2') ",
+                tableName));
+
+        QueryResult result = onTrino().executeQuery(format("SELECT unionLevel0.field0 FROM %s WHERE unionLevel0.field0 IS NOT NULL", tableName));
+        assertEquals(result.column(1).size(), 1);
```

Comment:
Shouldn't the size be 2 right?

Reply:
no, because one of the row don't have unionLevel0.field0, has unionLevel0.field1 instead.